libcontainer/cgroups/systemd/user.go

```diff
@@ -0,0 +1,86 @@
+// +build linux
+
+package systemd
+
+import (
+	"bufio"
+	"bytes"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strconv"
+	"strings"
+
+	systemdDbus "github.com/coreos/go-systemd/v22/dbus"
+	dbus "github.com/godbus/dbus/v5"
+	"github.com/opencontainers/runc/libcontainer/system"
+	"github.com/pkg/errors"
+)
+
+func newUserSystemdDbus() (*systemdDbus.Conn, error) {
+	addr, err := detectUserDbusSessionBusAddress()
+	if err != nil {
+		return nil, err
+	}
+	uid, err := detectUID()
+	if err != nil {
+		return nil, err
+	}
+
+	return systemdDbus.NewConnection(func() (*dbus.Conn, error) {
+		conn, err := dbus.Dial(addr)
+		if err != nil {
+			return nil, err
+		}
+		methods := []dbus.Auth{dbus.AuthExternal(strconv.Itoa(uid))}
+		err = conn.Auth(methods)
+		if err != nil {
+			conn.Close()
+			return nil, err
+		}
+		if err = conn.Hello(); err != nil {
+			conn.Close()
+			return nil, err
+		}
+		return conn, nil
+	})
+}
+
+// detectUID detects UID from $XDG_RUNTIME_DIR if running in userNS.
+// Otherwise returns os.Getuid() .
+func detectUID() (int, error) {
+	if !system.RunningInUserNS() {
+		return os.Getuid(), nil
+	}
+	// xdr is like "/run/user/1000" when the UID in the initial userNS is 1000
+	xdr := os.Getenv("XDG_RUNTIME_DIR")
+	if xdr == "" {
+		return -1, errors.New("could not detect the UID from XDG_RUNTIME_DIR")
+	}
+	s := strings.TrimPrefix(filepath.Clean(xdr), "/run/user/")
+	i, err := strconv.Atoi(s)
+	if err != nil {
+		return -1, errors.Wrap(err, "could not detect the UID from XDG_RUNTIME_DIR")
+	}
+	return i, nil
+}
```

Comment:
how will it work with something like `XDG_RUNTIME_DIR=/tmp/foo`?

Reply:
Who will set such value for systemd-mode?

Reply:
Maybe it's slightly better to assume that XDG_RUNTIME_DIR ends up with a slash and a user id, i.e. something like

```go
s := path.Base(xdr)
```

That would still be an assumption, but less strict that the path must be `/run/user/$ID`

Reply:
Let's discuss this in https://github.com/opencontainers/runc/pull/2281#discussion_r401922044